File: Data Structures & Algorithms/valid-tree/submission-4.py

class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) > n - 1:
            return False
        par = [i for i in range(n+1)]
        rank = [1] * (n+1)
        self.comp = n
        
        def find(n):
            if n != par[n]:
                par[n] = find(par[n])
            return par[n]

        def union(u, v):
            p1 = find(u)
            p2 = find(v)

            if p1 == p2:
                return False

            self.comp -= 1

            if rank[p1] > rank[p2]:
                p1, p2 = p2, p1
            rank[p1] += rank[p2]
            par[p2] = p1
            return True

        for u, v in edges:
            if not union(u, v):
                return False
        return self.comp == 1

File: Data Structures & Algorithms/redundant-connection/submission-4.cpp
class Solution {
public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size() + 1;
        vector<int> rank (n, 1);
        vector<int> par;
        for (int i = 0; i < n; i++) {
            par.push_back(i);
        }

        for (const auto &edge : edges) {
            if (!Union(edge[0], edge[1], par, rank)) {
                return vector<int>({edge[0], edge[1]});
            }
        }
        return {};
    }
private:
    int Find(int n, vector<int> &par) {
        if (n != par[n]) {
            par[n] = Find(par[n], par);
        }
        return par[n];
    }

    bool Union(int n1, int n2, vector<int> &par, vector<int> &rank) {
        int p1 = Find(n1, par);
        int p2 = Find(n2, par);

        if (p1 == p2) return false;

        if (rank[p1] < rank[p2]) {
            int temp = p1;
            p1 = p2;
            p2 = temp;
        }
        rank[p1] += rank[p2];
        par[p2] = p1;
        return true;
    }
};
